engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -102,19 +104,46 @@ public void setTailStatus(TailStatus tailStatus) {
   public abstract Object execute(
       VirtualFrame frame, State state, UnresolvedSymbol symbol, Object self, Object[] arguments);
 
-  @Specialization(guards = {"dispatch.hasType(self)", "!dispatch.hasSpecialDispatch(self)"})
+  @Specialization(guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)"})
   Object doFunctionalDispatch(
       VirtualFrame frame,
       State state,
       UnresolvedSymbol symbol,
       Object self,
       Object[] arguments,
-      @CachedLibrary(limit = "10") TypesLibrary dispatch,
+      @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
       @Cached MethodResolverNode methodResolverNode) {
-    Function function = methodResolverNode.expectNonNull(self, dispatch.getType(self), symbol);
+
+    Type selfTpe = typesLibrary.getType(self);
+    Function function = methodResolverNode.expectNonNull(self, selfTpe, symbol);
+
+    RootNode where = function.getCallTarget().getRootNode();
+    // If both Any and the type where `function` is declared, define `symbol`
+    // and the method is invoked statically, i.e. type of self is the eigentype,
+    // then we want to disambiguate method resolution by always resolved to the one in Any.
+    if (where instanceof MethodRootNode node && typeCanOverride(node, EnsoContext.get(this))) {
+      Function anyFun = symbol.getScope().lookupMethodDefinition(EnsoContext.get(this).getBuiltins().any(), symbol.getName());
```

Comment:
Alternatively, we could resolve `symbol` on `Any` first and then on `selfTpe`. That unfortunately leads to more complicated checks (one problematic example is for example `is_a` that is defined on `Any`, `Error`, `Meta` and who knows what else) so I didn't pursue it further.